src/optimization/ControlFlow.cpp

```diff
@@ -971,3 +974,87 @@ void optimizations::addStartStopSegment(const Module& module, Method& method, co
 
 	generateStopSegment(method);
 }
+
+LoopInclusionTreeNode* findRoot(LoopInclusionTreeNode *node)
+{
+	for (auto &parent : node->getNeighbors())
+	{
+		if (!parent.second.includes) {
+			// The root node must be only one
+			return findRoot(parent.first);
+		}
+	}
+	// this is root
+	return node;
+}
+using LoopInclusionTree = Graph<ControlFlowLoop*, LoopInclusionTreeNode>;
+
+void optimizations::removeConstantLoadInLoops(const Module& module, Method& method, const Configuration& config)
+{
+	// 1. find loops
+	auto cfg = ControlFlowGraph::createCFG(method);
+	auto loops = cfg.findLoops();
+
+	// 2. generate inclusion relation of loops as trees
+	LoopInclusionTree inclusionTree;
+	for (auto &loop1 : loops)
+	{
+		for (auto &loop2 : loops)
+		{
+			if (loop1 == loop2) continue;
+
+			if (loop1.includes(loop2))
+			{
+				auto &node1 = inclusionTree.getOrCreateNode(&loop1);
+				auto &node2 = inclusionTree.getOrCreateNode(&loop2);
+				node1.addNeighbor(&node2, LoopInclusion(true));
+				node2.addNeighbor(&node1, LoopInclusion(false));
+			}
+		}
+	}
+
+	logging::debug() << "loops" << logging::endl;
+	for (auto &loop : loops) {
+		logging::debug() << "  " << &loop << logging::endl;
```

Comment:
Printing the addresses of the loops found has no value and should be removed/rewritten!